### PR TITLE
[AIRFLOW-YYYY] WIP Rewrite Dataproc operators

### PR DESCRIPTION
**Pull request by nuclearpinguin**
# FYI:
- Create cluster operator has `PendingDeprecationWarning` to inform user about passing cluster objected instead of its parameters. Moreover the message has note about using `generate_cluster` method that should help user to migrate to new approach.
- Job operators has `DeprecationWarning` because user should use `DataprocSubmitJobOperator`. All operators implement `generate_job` method for easier migration.

In example dag you can see jobs an cluster generated using helper methods.

### CODE REVIEW

File: airflow/gcp/operators/dataproc.py [hunk 1]
```diff
@@ -576,126 +516,136 @@ class DataprocClusterScaleOperator(DataprocOperationBaseOperator):
...
-                    'numInstances': self.num_preemptible_workers
+                'secondary_worker_config': {
+                    'num_instances': self.num_preemptible_workers
                 }
             }
         }
         return scale_data
 
-    @staticmethod
-    def _get_graceful_decommission_timeout(timeout):
```
Comment (nuclearpinguin) [inline airflow/gcp/operators/dataproc.py]: This and optional_arguments were never used.

**Comment by nuclearpinguin** [comment]
System tests are working:
```
root@35fe4f73aab5:/workspace# nosetests tests.gcp.operators.test_dataproc_system -v -s
test_run_example_dag (tests.gcp.operators.test_dataproc_system.DataprocExampleDagsTest) ... [2019-09-25 14:58:52,324] {base_gcp_system_test_case.py:198} INFO - Storing DAGS from /workspace/dags to temporary directory /tmp/tmp__qbry9h
...
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/root/.virtualenvs/airflow36/lib/python3.6/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 14:59:10,413] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 14:59:10,414] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 14:59:10,612] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.create_cluster 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 14:59:18,104] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 14:59:18,105] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 14:59:18,306] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.scale_cluster 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 14:59:25,869] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 14:59:25,870] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 14:59:26,062] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.pig_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 14:59:48,211] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 14:59:48,212] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 14:59:48,406] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.spark_sql_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 15:00:25,641] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 15:00:25,642] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 15:00:25,855] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.spark_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 15:01:03,071] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 15:01:03,072] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 15:01:03,262] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.pyspark_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 15:01:40,441] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 15:01:40,442] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 15:01:40,633] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.hive_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 15:02:02,883] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 15:02:02,884] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 15:02:03,099] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.hadoop_task 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
/workspace/airflow/models/dagbag.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
[2019-09-25 15:02:55,451] {__init__.py:52} INFO - Using executor SequentialExecutor
[2019-09-25 15:02:55,452] {dagbag.py:90} INFO - Filling up the DagBag from /workspace/dags/example_dataproc.py
/workspace/airflow/gcp/operators/dataproc.py:215: PendingDeprecationWarning: Passing cluster parameters by keywords to `DataprocClusterCreateOperator` will be deprecated. Please provide cluster object using `cluster` parameter. You can use `generate_cluster` method to obtain cluster object.
  PendingDeprecationWarning, stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:843: DeprecationWarning: The `DataProcPigOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPigOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:975: DeprecationWarning: The `DataProcSparkSqlOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkSqlOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1050: DeprecationWarning: The `DataProcSparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcSparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1223: DeprecationWarning: The `DataProcPySparkOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcPySparkOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:909: DeprecationWarning: The `DataProcHiveOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHiveOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
/workspace/airflow/gcp/operators/dataproc.py:1123: DeprecationWarning: The `DataProcHadoopOperator` operator is deprecated, please use `DataprocSubmitJobOperator` instead. You can use `generate_job` method of `DataProcHadoopOperator` to generate dictionary representing your job and use it with the new operator.
  stacklevel=1
[2019-09-25 15:02:55,656] {cli.py:564} INFO - Running <TaskInstance: example_gcp_dataproc.delete_cluster 2019-09-24T00:00:00+00:00 [queued]> on host 35fe4f73aab5
Updated property [core/account].
ok

----------------------------------------------------------------------
Ran 1 test in 351.145s

OK
```

**Comment by nuclearpinguin** [comment]
@mik-laj @TobKed @potiuk your review would be welcomed. After that I will refactor tests.

File: airflow/gcp/hooks/dataproc.py [hunk 2]
```diff
@@ -366,126 +211,623 @@ def build(self) -> Dict:
...
+        :type metadata: Sequence[Tuple[str, str]]
+        """
+        client = self.get_cluster_client(location=region)
+        result = client.update_cluster(
+            project_id=project_id,
+            region=region,
+            cluster_name=cluster_name,
+            cluster=cluster,
+            update_mask=update_mask,
+            graceful_decommission_timeout=graceful_decommission_timeout,
```
Comment (potiuk) [inline airflow/gcp/hooks/dataproc.py]: Is graceful_decomission_timeout still used or deprecated ?

File: airflow/gcp/operators/dataproc.py (hunk 1 above)
Comment (nuclearpinguin) [reply airflow/gcp/operators/dataproc.py]: Good point! I missed the fact that those params were passed using `**self.optional_arguments` ✅

File: airflow/gcp/hooks/dataproc.py [hunk 3]
```diff
@@ -498,215 +840,59 @@ def submit(
...
+            job=job
+        )
+        job_id = job_object.reference.job_id
+        self.wait_for_job(
+            job_id=job_id,
+            location=region,
+            project_id=project_id
         )
 
-    def wait(self, operation: Dict) -> None:
```
[the remark this answers is not in the document]
Comment (nuclearpinguin) [reply airflow/gcp/hooks/dataproc.py]: Yes, I want to replace `wait_for_operation_done_or_error` and `wait_for_operation_done`with `OperationFutue.result()` pooling and `wait_for_job` to check submitted job as python client does not treat this as a long running operation.

File: airflow/gcp/operators/dataproc.py [hunk 4]
```diff
@@ -190,14 +161,16 @@ class DataprocClusterCreateOperator(DataprocOperationBaseOperator):
...
     def __init__(self,
-                 project_id: str,
-                 cluster_name: str,
-                 num_workers: int,
+                 project_id: Optional[str] = None,
+                 cluster: Optional[Dict] = None,
+                 cluster_name: Optional[str] = None,
+                 num_workers: Optional[int] = None,
+                 gcp_conn_id: str = "google_cloud_default",
                  zone: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (nuclearpinguin) [reply airflow/gcp/operators/dataproc.py]: As discussed I extracted this logic to `ClusterGenerator`.

File: airflow/gcp/hooks/dataproc.py (hunk 2 above)
Comment (nuclearpinguin) [reply airflow/gcp/hooks/dataproc.py]: Check, it's still in use, my mistake.

File: airflow/gcp/hooks/dataproc.py [hunk 5]
```diff
@@ -474,7 +474,7 @@ def list_clusters(
     @GoogleCloudBaseHook.fallback_to_default_project_id
     def update_cluster(  # pylint:disable=too-many-arguments
         self,
-        region: str,
```
Comment (nuclearpinguin) [inline airflow/gcp/hooks/dataproc.py]: This's a new method so it's backward compatible.

File: airflow/gcp/operators/dataproc.py [hunk 6]
```diff
@@ -1666,7 +1660,7 @@ class DataprocUpdateClusterOperator(BaseOperator):
     @apply_defaults
     def __init__(  # pylint:disable=too-many-arguments
         self,
-        region: str,
```
Comment (nuclearpinguin) [inline airflow/gcp/operators/dataproc.py]: This's a new operator so it's backward compatible.

File: airflow/gcp/hooks/dataproc.py [hunk 7]
```diff
@@ -366,126 +204,626 @@ def build(self) -> Dict:
...
+            metadata=metadata
+        )
 
     def submit(
         self,
         project_id: str,
         job: Dict,
         region: str = 'global',
-        job_error_states: Optional[Iterable[str]] = None
+        job_error_states: Optional[Iterable[str]] = None  # pylint:disable=unused-argument
```
Comment (TobKed) [inline airflow/gcp/hooks/dataproc.py]: Why this variable is not used?

File: airflow/gcp/hooks/dataproc.py [hunk 8]
```diff
@@ -23,209 +23,47 @@
...
 
-from googleapiclient.discovery import build
-from zope.deprecation import deprecation
+from cached_property import cached_property
+from google.api_core.retry import Retry
+from google.cloud.dataproc_v1beta2 import (  # pylint:disable=no-name-in-module
+    ClusterControllerClient, JobControllerClient, WorkflowTemplateServiceClient,
+)
+from google.cloud.dataproc_v1beta2.types import (  # pylint:disable=no-name-in-module
+    Cluster, Duration, FieldMask, Job, JobStatus, WorkflowTemplate,
```
Comment (TobKed) [inline airflow/gcp/hooks/dataproc.py]: I think `JobStatus` should lead to `enums`, I am not sure about the rest of the types.
There is a lot of mistakes in the `google.cloud` documentation with regard to `types` and `enums`:
https://github.com/PolideaInternal/airflow/pull/279#discussion_r320432094

https://github.com/googleapis/google-cloud-python/issues/7503
https://github.com/googleapis/google-cloud-python/pull/8779
https://github.com/googleapis/google-cloud-python/pull/9165

Comment (nuclearpinguin) [reply airflow/gcp/hooks/dataproc.py]: I prefere `types` as this is how documentation describe it, so it's easier to connect docs and code. Not everyone knows about enums and problems you mentioned.

File: airflow/gcp/hooks/dataproc.py (hunk 7 above)
Comment (nuclearpinguin) [reply airflow/gcp/hooks/dataproc.py]: It's here to preserve old function interface.

**Pull request closed without merging**